catalog/async/catalog-async-data/src/main/java/org/codice/ddf/catalog/async/data/impl/LazyProcessResourceImpl.java

```diff
@@ -0,0 +1,171 @@
+/**
+ * Copyright (c) Codice Foundation
+ * <p>
+ * This is free software: you can redistribute it and/or modify it under the terms of the GNU Lesser
+ * General Public License as published by the Free Software Foundation, either version 3 of the
+ * License, or any later version.
+ * <p>
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ddf.catalog.async.data.impl;
+
+import static org.codice.ddf.catalog.async.data.impl.ProcessResourceImpl.CONTENT_SCHEME;
+import static org.codice.ddf.catalog.async.data.impl.ProcessResourceImpl.DEFAULT_MIME_TYPE;
+import static org.codice.ddf.catalog.async.data.impl.ProcessResourceImpl.DEFAULT_NAME;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.function.Supplier;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang.StringUtils;
+import org.codice.ddf.catalog.async.data.api.internal.ProcessResource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import ddf.catalog.resource.Resource;
+
+public class LazyProcessResourceImpl implements ProcessResource {
+    private static final Logger LOGGER = LoggerFactory.getLogger(LazyProcessResourceImpl.class);
+
+    private URI uri;
+
+    private String name;
+
+    private String mimeType;
+
+    private long size;
+
+    private InputStream inputStream;
+
+    private String qualifier;
+
+    private boolean isModified = false;
+
+    private boolean isResourceLoaded = false;
+
+    private Supplier<Resource> resourceSupplier;
+
+    private String metacardId;
+
+    public LazyProcessResourceImpl(String metacardId, Supplier<Resource> resourceSupplier) {
+        if (StringUtils.isBlank(metacardId)) {
+            throw new IllegalArgumentException(
+                    "ProcessResourceImpl argument \"metacardId\" may not be null or empty.");
+        }
+        if (resourceSupplier == null) {
+            throw new IllegalArgumentException(
+                    "LazyProcessResourceImpl must have a non null resource supplier");
+        }
+
+        this.isModified = false;
+        this.metacardId = metacardId;
+        this.resourceSupplier = resourceSupplier;
+        this.isResourceLoaded = false;
+    }
+
+    @Override
+    public URI getUri() {
+        loadResource();
```

Comment:
❗️ Since a `ProcessResource` by definition can be remote or local, I wouldn't be opposed to giving control of loading the resource at the API level by adding `loadResource()` to the `ProcessResource` interface which must be called before interacting with the resource. And it would allow us to avoid situations of calling load like this all over. We could also throw an exception if we fail to load the resource in order to give the caller better error control.

Reply:
Leaving loadResource out of the interface. The ProcessResourceImpl assumes you already have the resource so this would only really affect the LazyProcessResourceImpl. If there was a loadResource call that needed to be made before interacting with the resource, the resource getter would probably still have to check if the resource had been loaded.